### ISSUE TITLE
OverflowError: Python int too large to convert to C long

### ISSUE THREAD

**Issue by cgohlke** [PyTables/PyTables#494]
Using PyTables 3.2.1 with NumPy 1.10.0b1 on win-amd64, I get 28 test errors of the kind `OverflowError: Python int too large to convert to C long` at https://github.com/PyTables/PyTables/blob/release/3.2.1/tables/tableextension.pyx#L828. I think this could be due to numpy 1.10.0b1 having relaxed strides enabled and arrays which have a dimension of size 1 get a stride of NPY_MAX_INTP. The real issue is that pytables often uses C `long` types for offsets, indices and strides. Those should be of type `ssize_t`. The following patch fixes the test failures, however there are other uses of `long`:
```patch
diff --git a/tables/tableextension.pyx b/tables/tableextension.pyx
index eae310a..cf11f50 100644
--- a/tables/tableextension.pyx
+++ b/tables/tableextension.pyx
@@ -699,7 +699,7 @@ cdef class Row:

   """

-  cdef long _row, _unsaved_nrows, _mod_nrows
+  cdef ssize_t _row, _unsaved_nrows, _mod_nrows, _stride, _rowsize
   cdef hsize_t start, absstep
   cdef long long stop, step, nextelement, _nrow, stopb  # has to be long long, not hsize_t, for neg
   cdef hsize_t nrowsinbuf, nrows, nrowsread
@@ -708,7 +708,7 @@ cdef class Row:
   cdef long long indexchunk
   cdef int     bufcounter, counter
   cdef int     exist_enum_cols
-  cdef int     _riterator, _stride, _rowsize, _write_to_seqcache
+  cdef int     _riterator, _write_to_seqcache
   cdef int     wherecond, indexed
   cdef int     ro_filemode, chunked
   cdef int     _bufferinfo_done, sss_on
```

```
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=
PyTables version:    3.2.1
HDF5 version:        1.8.15-patch1
NumPy version:       1.10.0b1
Numexpr version:     2.4.3 (using VML/MKL 11.3.0)
Zlib version:        1.2.8 (in Python interpreter)
LZO version:         2.09 (Feb 04 2015)
BZIP2 version:       1.0.6 (6-Sept-2010)
Blosc version:       1.4.4 (2015-05-05)
Blosc compressors:   blosclz (1.0.2.1), lz4 (1.7.1), lz4hc (1.7.1), snappy (1.1.1), zlib (1.2.8)
Cython version:      0.23.1
Python version:      3.5.0rc2 (v3.5.0rc2:cc15d736d860, Aug 25 2015, 05:08:37) [MSC v.1900 64 bit (AMD64)]
Platform:            Windows-10.0.10240
Byte-ordering:       little
Detected cores:      8
Default encoding:    utf-8
Default FS encoding: mbcs
Default locale:      (en_US, cp1252)
-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=

<snip>

======================================================================
ERROR: None (tables.tests.test_tables.SetItemTestCase1)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python35\lib\site-packages\tables\tests\test_tables.py", line 2993, in test06b
    for row in table.iterrows():
  File "X:\Python35\lib\site-packages\tables\table.py", line 1789, in iterrows
    row = tableextension.Row(self)
  File "tables\tableextension.pyx", line 787, in tables.tableextension.Row.__cinit__ (tables\tableextension.c:8925)
  File "tables\tableextension.pyx", line 828, in tables.tableextension.Row._new_buffer (tables\tableextension.c:9641)
OverflowError: Python int too large to convert to C long

======================================================================
ERROR: None (tables.tests.test_tables.SetItemTestCase1)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python35\lib\site-packages\tables\tests\test_tables.py", line 3029, in test07
    table.cols.col1[1:4:2] = [2, 3]
  File "X:\Python35\lib\site-packages\tables\table.py", line 3602, in __setitem__
    value, self.pathname)
  File "X:\Python35\lib\site-packages\tables\table.py", line 2460, in modify_column
    mod_recarr = self._read(start, stop, step)
  File "X:\Python35\lib\site-packages\tables\table.py", line 1887, in _read
    self.row._fill_col(result, start, stop, step, field)
  File "X:\Python35\lib\site-packages\tables\utils.py", line 244, in newfget
    mydict[name] = value = fget(self)
  File "X:\Python35\lib\site-packages\tables\table.py", line 598, in row
    return tableextension.Row(self)
  File "tables\tableextension.pyx", line 787, in tables.tableextension.Row.__cinit__ (tables\tableextension.c:8925)
  File "tables\tableextension.pyx", line 828, in tables.tableextension.Row._new_buffer (tables\tableextension.c:9641)
OverflowError: Python int too large to convert to C long

<snip>

======================================================================
ERROR: None (tables.tests.test_tablesMD.UpdateRowTestCase1)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "X:\Python35\lib\site-packages\tables\tests\test_tablesMD.py", line 2163, in test09b
    row = table.row
  File "X:\Python35\lib\site-packages\tables\utils.py", line 244, in newfget
    mydict[name] = value = fget(self)
  File "X:\Python35\lib\site-packages\tables\table.py", line 598, in row
    return tableextension.Row(self)
  File "tables\tableextension.pyx", line 787, in tables.tableextension.Row.__cinit__ (tables\tableextension.c:8925)
  File "tables\tableextension.pyx", line 828, in tables.tableextension.Row._new_buffer (tables\tableextension.c:9641)
OverflowError: Python int too large to convert to C long

----------------------------------------------------------------------
Ran 5737 tests in 252.000s

FAILED (errors=29, skipped=56)
```

**Comment by andreabedini** [comment]
Thanks @cgohlke I'll look into this soon

**Comment by andreabedini** [comment]
@cgohlke shouldn't it be `npy_intp` the right type for offsets, indices and strides?

**Comment by cgohlke** [comment]
Numpy uses `npy_intp` / `Py_intptr_t` for offsets, indices and strides of arrays. Python uses `ssize_t` / `Py_ssize_t`. For all practical purposes it should not matter [1]. Except that Cython has a bug [2] where `npy_intp` indices are sometimes temporarily stored in C `long` variables.
[1] https://www.python.org/dev/peps/pep-0353/#why-not-py-intptr-t
[2] https://github.com/scipy/scipy/pull/5008

**Comment by andreabedini** [comment]
Making `_stride` a `npy_intp` works for me. `rowsize` is set from `dtype.itemsize` which is a simple `int`.

**Comment by cgohlke** [comment]
> `dtype.itemsize` which is a simple `int`

That's another flaw (https://github.com/numpy/numpy/issues/6235)

**Comment by andreabedini** [comment]
Sorry for the accidental close. I'm on my phone. What is the flaw there? The issue you link to is about a missing overflow check.

**Comment by andreabedini** [comment]
This should be fixed in a675a25.
